Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = DateTime.UtcNow,
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds =
+                    options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = DateTime.MaxValue - doc.LastRefresh.Value;
+                doc.SlidingExpirationSeconds = forever.TotalSeconds / 2;
+            }
+            if (options.AbsoluteExpiration.HasValue)
+            {
+                doc.AbsoluteExpiration = options.AbsoluteExpiration.Value.UtcDateTime;
+            }
+            if (options.AbsoluteExpirationRelativeToNow.HasValue)
+            {
+                doc.AbsoluteExpiration = DateTime.UtcNow +
+                    options.AbsoluteExpirationRelativeToNow.Value;
+            }
+            return doc;
+        }
+
+        void IDistributedCache.Set(string key, byte[] value, DistributedCacheEntryOptions options) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options)).Wait();
+
+        Task IDistributedCache.SetAsync(string key, byte[] value, DistributedCacheEntryOptions options,
+            CancellationToken token) =>
+            _cacheEntries.Document(key).SetAsync(MakeCacheDoc(value, options),
+            cancellationToken: token);
+
+        /// <summary>
+        /// Scans the firestore collection for expired cache entries and
+        /// deletes them.  
+        /// </summary>
+        /// <param name="token">A cancellation token.</param>
+        public async Task CollectGarbageAsync(CancellationToken token)
+        {
+            _logger.LogTrace("Begin garbage collection.");
+            // Purge entries whose AbsoluteExpiration has passed.
+            const int pageSize = 40;
+            int batchSize;  // Batches of cache entries to be deleted.
+            var now = DateTime.UtcNow;
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderByDescending("AbsoluteExpiration")
+                    .StartAfter(now)
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    if (docSnapshot.ConvertTo<CacheDoc>()
+                        .AbsoluteExpiration.HasValue)
+                    {
+                        writeBatch.Delete(docSnapshot.Reference,

```

Comment:
Guidance from the Firestore team is just to call `DeleteAsync` on the document. The limits for batching aren't as well-defined as we might like.

Reply:
One at a time?